### PR TITLE
feat(pt_expt): multi-rank LAMMPS support for GNN models (DPA3 / DPA2 / spin)

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary

Adds multi-rank LAMMPS support to the `pt_expt` (.pt2 / AOTInductor) backend for GNN models — DPA3 (repflows) and DPA2 (repformers), plus spin GNN — at parity with the existing `pt` (.pth / torch.jit) backend. Without this, multi-rank LAMMPS users with GNN .pt2 models fall back to single-rank-only, and the C++ side crashes on the first ghost exchange when given a non-`use_loc_mapping` GNN .pt2.

The mechanism mirrors the pt backend's per-layer ghost-atom MPI exchange: each repflow/repformer block exchanges `g1` across ranks via `border_op` so each rank sees up-to-date ghost embeddings. To survive `torch.export` + AOTInductor packaging, `border_op` is wrapped as an opaque `torch.library.custom_op` (`deepmd_export::border_op`) with a separate `border_op_backward` C++ symbol for autograd.

## Design

- **Phase 1 — dpmodel plumbing**: thread `comm_dict: dict | None = None` through `make_model`, `base_atomic_model`, descriptor wrappers (dpa1/dpa2/dpa3/hybrid/se_*), and the repflows/repformers blocks. Lift the per-layer `node_ebd_ext` construction into a `_exchange_ghosts` method (default array-api impl ignores `comm_dict`).
- **Phase 2 — pt_expt opaque op + block overrides**:
  - `deepmd::border_op_backward` C++ op (additive accumulation into local atom slots — symmetric exchange used by autograd backward).
  - `deepmd_export::border_op` Python `custom_op` wrapper with `register_fake` and `register_autograd` so the op is opaque to `torch.export`.
  - `pt_expt/descriptor/repflows.py` and `repformers.py` block subclasses with `_exchange_ghosts` overrides that call the opaque op (with the spin real/virtual split + `concat_switch_virtual` when `comm_dict[\"has_spin\"]` is set).
- **Phase 3 — dual-artifact AOTInductor export**: when `_has_message_passing(model)` is true, compile **two** artifacts into the .pt2 ZIP:
  - `forward_lower_no_comm.pt2` — current behavior (single-rank, mapping-based gather).
  - `forward_lower_with_comm.pt2` — adds positional comm tensors `(send_list, send_proc, recv_proc, send_num, recv_num, communicator, nlocal, nghost)` to the trace input list, plus `has_spin=tensor([1])` baked in for spin GNN.
  - Metadata: `has_message_passing` + `has_comm_artifact` flags so the C++ loader picks the right artifact.
- **Phase 4 — C++ dispatch**: `DeepPotPTExpt::compute` and `DeepSpinPTExpt::compute` route to the with-comm artifact when `lmp_list.nswap > 0`. `commPTExpt` adds `build_comm_tensors_positional` and `build_comm_tensors_positional_with_virtual_atoms` (the latter remaps sendlists via `fwd_map` when NULL atoms drop out of the model's view).
- **Phase 5 — LAMMPS tests**: end-to-end multi-rank tests (DPA3, DPA2, spin DPA3) covering the basic case + the structural edge cases (NULL-type atoms, empty subdomain, all-NULL rank, isolated NULL, nlist rebuild, N>2 decomposition).

## Coverage matrix

|  | DPA3 | DPA2 | spin DPA3 |
| - | - | - | - |
| basic mpi-2 | ✓ | ✓ | ✓ |
| empty subdomain (rank 1 empty) | ✓ | — | ✓ |
| all-NULL rank | ✓ | — | — |
| isolated NULL | ✓ | — | — |
| NULL atoms straddling boundary | ✓ | — | ✓ |
| NULL across nlist rebuild | ✓ | — | — |
| N>2 (2x2x1, 4x1x1, 2x2x2) | ✓ | — | — |
| cached mapping_tensor (ago>0) | ✓ | — | — |

Tests compare mpi-N vs same-archive mpi-1 for force / force_mag / virial / energy (atol 1e-8); no hardcoded numerical references.

Plus a unit test (`test_has_message_passing.py`) pinning the `_has_message_passing` schema-drift contract.

## Co-existence with #5407

This branch was rebased onto upstream master after PR #5407 (.pt2 perf) merged. The merge required:
- `forward_common_lower_exportable_with_comm` (spin and non-spin variants) now applies `_pad_nlist_for_export` + `need_sorted_nlist_for_lower=True` workarounds matching #5407's regular variant — keeps the with-comm trace's `nnei` axis dynamic.
- CUDA `realize_opcount_threshold=0` workaround applied around BOTH artifact compiles.
- `do_atomic_virial=True` is used for all multi-rank fixtures to avoid AOTI compile-time changes from #5407's default `=False`. (Multi-rank with `do_atomic_virial=False` is a known coverage gap — see Limitations.)

## Known limitations

- **Hybrid + GNN + multi-rank**: `_has_message_passing` doesn't recurse into hybrid children → no with-comm artifact produced for hybrid-of-GNN. Multi-rank LAMMPS with such a model would fall back to single-artifact and crash on first ghost exchange. Out of scope for this PR; if a user hits this, they get the same crash they had before.
- **`do_atomic_virial=False` under multi-rank**: production default. All multi-rank tests use `=True` (matches #5407 fixture conventions). Not exercised end-to-end yet.
- **CUDA**: the `TORCH_LIBRARY_IMPL(deepmd_export, CUDA, m)` registration exists and the symbol is callable, but no GPU end-to-end test runs in this PR (CPU-only build environment locally).
- **DPA2 (repformers) edge cases**: only basic mpi-2 is tested for DPA2. NULL / empty-subdomain / nlist-rebuild covered for DPA3 only — these paths are descriptor-agnostic by construction but not exercised end-to-end for DPA2.
- **Spin DPA3 edge cases**: empty subdomain and NULL straddling boundary covered. all-NULL-rank, isolated-NULL, nlist-rebuild not.
- **Multiple spin-active types** (`use_spin=[True, True, ...]`): only `[True, False]` tested.
- **Frame batching nb>1 with `comm_dict`**: `_exchange_ghosts` uses `.squeeze(0)` / `.unsqueeze(0)` (mirrors pt). LAMMPS feeds nb=1 — fine in practice; breaks if reused outside LAMMPS.
- **Float16/bfloat16**: comm_dict path was developed in float64 only.
- **Cross-backend (jax/paddle/numpy) `comm_dict=None` neutrality**: dpmodel default `_exchange_ghosts` is the original code lifted into a method, behaviorally equivalent. Not separately re-tested via running consistent tests with `comm_dict=None` explicitly threaded.
- **Old .pt2 files lacking `has_comm_artifact` metadata**: C++ defaults to single-artifact when key is missing. Not negative-tested.
- **AOTI compile failure cleanup**: if the with-comm compile fails after the regular artifact is written, the resulting .pt2 is half-formed. RAII / SIGKILL leakage of `TempFile` in /tmp also pre-exists.

A full catalog of touched-but-untested paths is maintained at `memory/gnn_mpi_untested_paths.md` (local to the author).

## Test plan

- [ ] `pytest source/tests/pt_expt/` — eager-parity, export round-trip, schema-drift unit tests
- [ ] `pytest source/tests/consistent/descriptor/test_dpa3.py source/tests/consistent/descriptor/test_dpa2.py` — non-regression for the single-rank path
- [ ] `pytest source/lmp/tests/test_lammps_dpa3_pt2.py source/lmp/tests/test_lammps_dpa2_pt2.py source/lmp/tests/test_lammps_spin_dpa3_pt2.py` — multi-rank end-to-end (requires mpirun + mpi4py)
- [ ] `ctest -R PtExpt` — C++ tests (single-rank, since multi-rank is exercised via LAMMPS)
- [ ] CI runs across CPU + CUDA matrix

## Summary by CodeRabbit

* **New Features**
  * Optional comms/MPI metadata support across inference paths enabling ghost-exchange; dual “with-comm” export/run artifact and tracing entrypoints for message-passing GNNs and spin models, plus improved export/tracing support for comm-enabled ops.

* **Bug Fixes**
  * Fixed unsafe neighbor-list memory handling.

* **Tests**
  * Added extensive MPI parity, export-with-comm, repflow/repformer parallel-path, spin integration tests and CLI MPI runner scripts.

### CODE REVIEW

File: deepmd/dpmodel/descriptor/hybrid.py [hunk 1]
```diff
@@ -332,7 +333,9 @@ def call(
                 # mixed_types is True, but descrpt.mixed_types is False
                 assert nl_distinguish_types is not None
                 nl = nl_distinguish_types[:, :, nci]
-            odescriptor, gr, g2, h2, sw = descrpt(coord_ext, atype_ext, nl, mapping)
+            odescriptor, gr, g2, h2, sw = descrpt(
+                coord_ext, atype_ext, nl, mapping, comm_dict=comm_dict
+            )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/descriptor/hybrid.py]: The unused-unpack rename (`g2/h2/sw` → `_g2/_h2/_sw`) was applied in commit `bf1685ffc`. The `fparam` forwarding suggestion is **valid but pre-existing**: `DescrptHybrid.call` has dropped `fparam` long before this PR — `git blame` on the enclosing function shows fparam was already accepted-but-unforwarded. This PR's refactor only added `comm_dict` to the same call site without changing the fparam handling. Addressing fparam forwarding requires a regression test for hybrid descriptors wrapping fparam-aware children, which is out of scope for the GNN MPI work. Tracking as a known pre-existing limitation; will be addressed in a follow-up PR.

File: deepmd/pt_expt/utils/serialization.py [hunk 2]
```diff
@@ -78,6 +79,103 @@ def _json_to_numpy(model_obj: dict) -> dict:
...
+    # Walk into the GNN block (repflows / repformers) to inspect
+    # ``use_loc_mapping``. The attribute lives on the block, not on the
+    # top-level descriptor wrapper.
+    for attr in ("repflows", "repformers"):
+        block = getattr(descriptor, attr, None)
+        if block is None:
+            continue
+        if getattr(block, "use_loc_mapping", False):
+            return False
+    return True
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/serialization.py]: Valid but **out of scope** for this PR. This is Tier-1 #1 in the project's `gnn_mpi_untested_paths.md` catalog of known limitations. Recursing `_has_message_passing` into hybrid wrappers requires a hybrid+GNN multi-rank fixture (so we can verify the with-comm artifact is actually correct for hybrid models, not just produced) plus end-to-end LAMMPS multi-rank testing of hybrid+GNN. Both are substantial work and will land in a separate PR.

For now the existing behaviour (silent fall-back to single-artifact) preserves the pre-PR status quo: hybrid descriptors with GNN children continue to work in single-rank mode just as they did before; multi-rank with this configuration was never supported and will fail loudly at the first ghost exchange via `border_op` if attempted.

File: source/lmp/tests/test_lammps_dpa3_pt2.py [hunk 3]
```diff
@@ -315,3 +415,489 @@ def test_pair_deepmd_si(lammps_si) -> None:
...
+        ]
+        if processors is not None:
+            argv.extend(["--processors", processors])
+        elif nprocs == 1:
+            argv.extend(["--processors", "1 1 1"])
+        if extra_args:
+            argv.extend(extra_args)
+        if runner_args:
+            argv.extend(runner_args)
+        sp.check_call(argv)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_dpa3_pt2.py:404]: Not applicable to this project. `pyproject.toml` does not enable any `S` (flake8-bandit) rules — the lint config selects `E,W,F,D,UP,C4,RUF,NPY,TID*,T20,B904,N804,N805,DTZ,TCH,PYI,ANN,B905`. `ruff check` passes cleanly without the annotation. Adding `# noqa: S603` would be a no-op comment for a rule the project does not run.

File: source/lmp/tests/test_lammps_spin_dpa3_pt2.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        ]
+        if processors is not None:
+            argv.extend(["--processors", processors])
+        elif nprocs == 1:
+            argv.extend(["--processors", "1 1 1"])
+        if extra_args:
+            argv.extend(extra_args)
+        if runner_args:
+            argv.extend(runner_args)
+        sp.check_call(argv)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_spin_dpa3_pt2.py:176]: Same as the analogous comment on `test_lammps_dpa3_pt2.py:479`: `pyproject.toml` does not enable any `S` (flake8-bandit) rules, so `ruff check` passes already and a `# noqa: S603` would be a no-op annotation.

File: source/tests/pt_expt/model/test_export_with_comm.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    (
+        send_list,
+        send_proc,
+        recv_proc,
+        send_num,
+        recv_num,
+        communicator,
+        nlocal,
+        nghost_t,
+    ) = comm_inputs
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/model/test_export_with_comm.py:231]: Not applicable to this project. `pyproject.toml` line 399 explicitly ignores `RUF059` (`# unused-unpacked-variable`), so the `communicator` binding here does not violate the project's lint policy and `ruff check` passes. Renaming to `_communicator` would be cosmetic only.

File: source/op/pt/comm.cc [hunk 6]
```diff
@@ -363,4 +418,152 @@ std::vector<torch::Tensor> border_op(const torch::Tensor& sendlist_tensor,
...
+                               const torch::Tensor& g1_tensor,
+                               const torch::Tensor& communicator_tensor,
+                               const torch::Tensor& nlocal_tensor,
+                               const torch::Tensor& nghost_tensor) {
+  auto out = border_op(sendlist_tensor, sendproc_tensor, recvproc_tensor,
+                       sendnum_tensor, recvnum_tensor, g1_tensor,
+                       communicator_tensor, nlocal_tensor, nghost_tensor);
+  // border_op returns {g1_tensor} — a list whose first element aliases
+  // g1_tensor. Clone for AOTI graph-output correctness.
+  return out.empty() ? torch::empty_like(g1_tensor) : out[0].clone();
```
Comment (iProzd) [inline source/op/pt/comm.cc]: `out.empty()` check masks internal errors. If `border_op` returns an empty list, returning `empty_like(g1_tensor)` produces silent zero outputs instead of surfacing the bug.

  Is it possible to do like this?
```
  if (out.empty()) {
    throw std::runtime_error("border_op returned empty output");
  }
  return out[0].clone();
```

File: source/api_cc/src/DeepPotPTExpt.cc [hunk 7]
```diff
@@ -154,6 +162,38 @@ void DeepPotPTExpt::init(const std::string& model,
...
+              with_comm_tempfile_->path(), "model", false, 1,
+              gpu_enabled ? static_cast<c10::DeviceIndex>(gpu_id)
+                          : static_cast<c10::DeviceIndex>(-1));
+    } catch (const std::exception& e) {
+      std::cerr << "DeepPotPTExpt: failed to load with-comm artifact ("
+                << e.what() << "); falling back to single-rank-only dispatch."
+                << std::endl;
+      with_comm_tempfile_.reset();
+      with_comm_loader.reset();
+      has_comm_artifact_ = false;
```
Comment (iProzd) [inline source/api_cc/src/DeepPotPTExpt.cc]: Silent fallback when with-comm artifact fails to load can cause incorrect results in multi-rank LAMMPS. 

If the artifact load fails but the user runs a GNN model with nswap>0, the code falls back to the regular artifact which doesn't update ghost embeddings — producing silently wrong results instead of failing fast.

  Suggest: throw when multi-rank is actually used:
```
  if (use_with_comm && !with_comm_loader) {
    throw deepmd::deepmd_exception(
        "Multi-rank LAMMPS requires with-comm artifact but it failed to load. "
        "See earlier error message.");
  }
```

**Review by iProzd — CHANGES_REQUESTED** [review]

**Review by iProzd — APPROVED** [review]
